groups/service.go

```diff
@@ -0,0 +1,415 @@
+				}
+				if as.group.Name != "" {
+					g.Name = as.group.Name
+				}
+				if as.group.Description != "" {
+					g.Description = as.group.Description
+				}
+				g4, err := as.client.InsertGroup(&g)
+				if err != nil {
+					return fmt.Errorf("unable to add new group %q: %v", as.group.EmailId, err)
+				}
+				log.Printf("> Successfully created group %s\n", g4.Email)
+			}
+		} else {
+			return fmt.Errorf("unable to fetch group %q: %#v", as.group.EmailId, err.Error())
+		}
+	} else {
+		if as.group.Name != "" && group.Name != as.group.Name ||
+			as.group.Description != "" && group.Description != as.group.Description {
+			if !config.ConfirmChanges {
+				log.Printf("dry-run: would update group name/description %q\n", as.group.EmailId)
+			} else {
+				log.Printf("Trying to update group: %q\n", as.group.EmailId)
+				g := admin.Group{
+					Email: as.group.EmailId,
+				}
+				if as.group.Name != "" {
+					g.Name = as.group.Name
+				}
+				if as.group.Description != "" {
+					g.Description = as.group.Description
+				}
+				g4, err := as.client.UpdateGroup(as.group.EmailId, &g)
+				if err != nil {
+					return fmt.Errorf("unable to update group %q: %v", as.group.EmailId, err)
+				}
+				log.Printf("> Successfully updated group %s\n", g4.Email)
+			}
+		}
+	}
+	return nil
+}
+
+func (as *adminService) DeleteGroupsIfNecessary() error {
+	g, err := as.client.ListGroups()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+
+	for _, g := range g.Groups {
+		found := false
+		for _, g2 := range groupsConfig.Groups {
+			if g2.EmailId == g.Email {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+
+		// We did not find the group in our groups.xml, so delete the group
+		if config.ConfirmChanges {
+			if *verbose {
+				log.Printf("deleting group %s", g.Email)
+			}
+			err := as.client.DeleteGroup(g.Email)
+			if err != nil {
+				return fmt.Errorf("unable to remove group %s : %v", g.Email, err)
+			}
+			log.Printf("Removing group %s\n", g.Email)
+		} else {
+			log.Printf("dry-run: would remove group %s\n", g.Email)
+		}
+
+	}
+
+	return nil
+}
+
+func (as *adminService) RemoveOwnerOrManagersFromGroup(members []string) error {
+	if *verbose {
+		log.Printf("removeOwnerOrManagersGroup %q %v", as.group.EmailId, members)
+	}
+	l, err := as.client.ListMembers(as.group.EmailId)
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			log.Printf("skipping removing members group %q as group has not yet been created\n", as.group.EmailId)
+			return nil
+		}
+		return fmt.Errorf("unable to retrieve members in group %q: %v", as.group.EmailId, err)
+	}
+
+	for _, m := range l.Members {
+		found := false
+		for _, m2 := range members {
+			if m2 == m.Email {
+				found = true
+				break
+			}
+		}
+		if found || m.Role == "MEMBER" {
```

Comment:
I know this is how you found it, but I scratched my head over this for a while.  Do you think you can add a comment explaining why this is needed? It's the only thing that makes `RemoveOwnersOrManagersFromGroup` different than `RemoveMembersFromGroup`

It would be nice to refactor this away as a followup.  I think this is to handle someone moving from manager to member, and not dropping them.  But making two passes over all group members seems like an inefficiency we'll want to remove

Reply:
> I think this is to handle someone moving from manager to member, and not dropping them.

If I understand correctly, this is along the lines of:

If a member m exists in our desired list of members, do nothing.
However, if this member m does not exist in our desired list of
members but is in the role of a MEMBER (non OWNER/MANAGER), still
do nothing since we care only about OWNER/MANAGER roles here.